python/pyarrow/array.pxi

```diff
@@ -1778,6 +1778,40 @@ cdef class Array(_PandasConvertible):
 
         return pyarrow_wrap_array(array)
 
+    def __dlpack__(self, stream=None):
+        """Export a primitive array as a DLPack capsule.
+
+        Parameters
+        ----------
+        stream : int, optional
+            A Python integer representing a pointer to a stream. Currently not supported.
+            Stream is provided by the consumer to the producer to instruct the producer
+            to ensure that operations can safely be performed on the array.
+
+        Returns
+        -------
+        capsule : PyCapsule
+            A DLPack capsule for the array, containing a DLPackManagedTensor.
+        """
+        if stream is None:
+            return to_dlpack(self)
+        else:
+            raise NotImplementedError(
+                "Only stream=None is supported."
+            )
+
+    def __dlpack_device__(self):
+        """
+        Performs the operation __dlpack_device__.
```

Comment:
Where is this specified? I see nothing in https://dmlc.github.io/dlpack/latest/python_spec.html

Reply:
There is exactly one mention of `__dlpack_device__` on that page (just the mention that it exists in bullet point no 2), but unfortunately without any explanation of what it is or should return. 

So implementation defined, eg https://github.com/numpy/numpy/blob/eabb962d38d8a871b2d76a9198846b49897365eb/numpy/_core/src/multiarray/dlpack.c#L279-L287 or https://github.com/pytorch/pytorch/blob/89569be2bdb608e64a190c5498f36b711ddd3e8d/torch/_tensor.py#L1484-L1502

Reply:
Ah, it seems this is somewhat document on the Array API side (not on the DLPack page itself): https://data-apis.org/array-api/latest/API_specification/generated/array_api.array.__dlpack_device__.html

Reply:
I have missed it too. It is in the second item in the enumerated list at `Syntax for data interchange with DLPack`:
https://dmlc.github.io/dlpack/latest/python_spec.html#syntax-for-data-interchange-with-dlpack

Numpy doesn't use it in their `from_dlpack(x)` but they have it implemented for the array object:
https://github.com/numpy/numpy/blob/eabb962d38d8a871b2d76a9198846b49897365eb/numpy/array_api/_array_object.py#L488-L493

PyTorch checks for `__dlpack_device__` in Python:
https://github.com/pytorch/pytorch/blob/89569be2bdb608e64a190c5498f36b711ddd3e8d/torch/utils/dlpack.py#L100-L101
but I am not sure if they have something similar in C++ (I think not).

Reply:
@AlenkaF have you checked it locally that this works with pytorch (or another tensor library that might check `__dlpack_device__` (ideally we would have integration tests for this part as well instead of only with numpy, but this alone is probably not worth adding a heavy test dependency on pytorch)

Reply:
Yes, I have (added to the docs examples - those are not being tested yet, right?)